app/assets/javascripts/govuk_publishing_components/analytics/auto-scroll-tracker.js

```diff
@@ -0,0 +1,184 @@
+        // store event listener functions as variables so they can be removed if needed
+        this.scrollEvent = this.onScroll.bind(this)
+        window.addEventListener('scroll', this.scrollEvent)
+        this.resizeEvent = this.onResize.bind(this)
+        window.addEventListener('resize', this.resizeEvent)
+
+        // check if the page height changes e.g. accordion opened
+        this.interval = window.setInterval(function () {
+          var pageHeight = document.querySelector('body').clientHeight
+          if (pageHeight !== this.pageHeight) {
+            this.pageHeight = pageHeight
+            this.getWindowDetails()
+            this.trackVisibleNodes()
+          }
+        }.bind(this), this.PAGEHEIGHT_TIMEOUT_DELAY)
+      }
+    }
+  }
+
+  AutoScrollTracker.prototype.onScroll = function () {
+    clearTimeout(this.scrollTimeout)
+    this.scrollTimeout = setTimeout(function () {
+      this.trackVisibleNodes()
+      this.trackVisibleNodes()
+    }.bind(this), this.SCROLL_TIMEOUT_DELAY)
+  }
+
+  AutoScrollTracker.prototype.onResize = function () {
+    clearTimeout(this.resizeTimeout)
+    this.resizeTimeout = setTimeout(function () {
+      this.getWindowDetails()
+      this.trackVisibleNodes()
+    }.bind(this), this.RESIZE_TIMEOUT_DELAY)
+  }
+
+  AutoScrollTracker.prototype.getWindowDetails = function () {
+    this.pageHeight = document.querySelector('body').clientHeight
+    this.windowHeight = window.innerHeight
+    this.trackedNodes = this.track.getTrackingNodes(this.trackedNodes)
+  }
+
+  AutoScrollTracker.prototype.trackVisibleNodes = function () {
+    for (var i = 0; i < this.trackedNodes.length; i++) {
+      var node = this.trackedNodes[i]
+      if (this.isVisible(node.top, node.bottom) && !node.alreadySeen) {
+        node.alreadySeen = true
+        // we store whether a heading has been tracked or not on the heading
+        // because if headings appear/disapper (e.g. inside an accordion)
+        // the order changes, so we can't refer to the previous trackedNodes
+        // as we do with percentages
+        if (node.node) {
+          node.node.setAttribute('data-autoscrolltracker-already-seen', true)
+        }
+
+        var action = node.eventData.action
+        var label = node.eventData.label
+
+        if (window.GOVUK.analytics && window.GOVUK.analytics.trackEvent) {
+          window.GOVUK.analytics.trackEvent('ScrollTo', action, { label: label, nonInteraction: true })
+        }
+      }
+    }
+  }
+
+  AutoScrollTracker.prototype.isVisible = function (top, bottom) {
+    return window.scrollY <= top && (window.scrollY + this.windowHeight) >= bottom
+  }
+
+  AutoScrollTracker.Heading = function (config) {
+    this.config = config
+  }
+
+  AutoScrollTracker.Heading.prototype.getTrackingNodes = function () {
+    var headingsDetails = []
+    var headings = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6']
+    var headingsFound = document.querySelectorAll(headings)
+    if (headingsFound) {
```

Comment:
I think this line is redundant, as you'll always get something that returns true and something you can iterate through from a `querySelectorAll`

Reply:
Thanks, didn't know that querySelectorAll returned an empty array on not found. Changed.